pkg/rpc/auth.go

```diff
@@ -92,134 +119,237 @@ func (a kvAuth) streamInterceptor(
+			//   the tenant ID in the cert.
+			if wantedTenantID.IsSet() {
+				// Verify conformance.
+				if tlsID != wantedTenantID {
+					return nil, authErrorf(
+						"client wants to authenticate as tenant %v, but is using TLS cert for tenant %v",
+						wantedTenantID, tlsID)
+				}
+			} else {
+				// Use the ID in the cert as wanted tenant ID.
+				wantedTenantID = tlsID
+			}
+		} else {
+			// We are using TLS, but the peer is not using a client tenant
+			// cert. In that case, we only allow RPCs if the principal is
+			// 'node' or 'root' and the tenant scope in the cert matches
+			// this server (either the cert has scope "global" or its scope
+			// tenant ID matches our own).
+			//
+			// TODO(benesch): the vast majority of RPCs should be limited to just
+			// NodeUser. This is not a security concern, as RootUser has access to
+			// read and write all data, merely good hygiene. For example, there is
+			// no reason to permit the root user to send raw Raft RPCs.
+			certUserScope, err := security.GetCertificateUserScope(clientCert)
+			if err != nil {
+				return nil, err
+			}
+			if err := checkRootOrNodeInScope(certUserScope, a.tenant.tenantID); err != nil {
+				return nil, err
+			}
 		}
 	}
 
-	// Here we handle the following cases:
-	//
-	// - incoming connection from a RPC admin client into either a KV
-	//   node or a SQL server, using a valid root or node client cert.
-	// - incoming connections from another KV node into a KV node, using
-	//   a node client cert.
-	// - calls coming through the gRPC gateway, from an HTTP client. The gRPC
-	//   gateway uses a connection dialed as the node user.
-	//
-	// In both cases, we must check that the client cert is either root
-	// or node. We also need to check that the tenant scope for the cert
-	// is either the system tenant ID or matches the tenant ID of the server.
-
-	// TODO(benesch): the vast majority of RPCs should be limited to just
-	// NodeUser. This is not a security concern, as RootUser has access to
-	// read and write all data, merely good hygiene. For example, there is
-	// no reason to permit the root user to send raw Raft RPCs.
-	certUserScope, err := security.GetCertificateUserScope(&tlsInfo.State)
-	if err != nil {
-		return roachpb.TenantID{}, err
+	if wantedTenantID.IsSet() {
+		// Consistency check: the peer wants to authenticate as another
+		// tenant server, verify that the peer is a service for the same
+		// tenant as ourselves (we don't want to allow tenant 123 to serve
+		// requests for a client coming from tenant 456) or we are the system tenant.
+		if !a.tenant.tenantID.IsSystem() && wantedTenantID != a.tenant.tenantID {
+			log.Ops.Infof(ctx, "rejected incoming request from tenant %d (misconfiguration?)", wantedTenantID)
+			return nil, authErrorf("client tenant identity (%v) does not match server", wantedTenantID)
+		}
+		return authnSuccessPeerIsTenantServer(wantedTenantID), nil
 	}
+	return authnSuccessPeerIsPrivileged{}, nil
```

Comment:
Some of the previous discussion is around it being hard to see what cases lead us to hit this case. I think part of the difficulty is tracking when wantedTenantID is set and unset. I wonder if a simple refactoring might help (warning: code written directly in browser).  I still don't love it, but it is nice to be able to directly scan for `return authnSuccessPeerIsPrivileged`. Anyway, perhaps this is just pushing peas around the plate.

```golang
// authenticate verifies the credentials of the client and performs
// some consistency check with the information provided.
func (a kvAuth) authenticate(ctx context.Context) (authnResult, error) {
	var ar authnResult
	if clientTenantID, localRequest := grpcutil.IsLocalRequestContext(ctx); localRequest {
		var err error
		ar, err = authenticateLocalRequest(ctx, clientTenantID)
		if err != nil {
			return nil, err
		}
	} else {
		var err error
		ar, err = authenticateNeworkRequest(ctx)
		if err != nil {
			return nil, err
		}
	}

	switch res := ar.(type) {
	case authnSuccessPeerIsTenantServer:
		if res != a.tenant.tenantID {
			log.Ops.Infof(ctx, "rejected incoming request from tenant %d (misconfiguration?)", wantedTenantID)
			return nil, authErrorf("client tenant identity (%v) does not match server", wantedTenantID)
		}
	case authnSuccessPeerIsPrivileged:
	default:
		return nil, errors.AssertionFailedf("programming error: unhandled case %T", ar)
	}

	return ar, nil

}

// Deal with local requests done through the
// internalClientAdapter. There's no TLS for these calls, so the
// regular authentication code path doesn't apply. The clientTenantID
// should be the result of a call to grpcutil.IsLocalRequestContext.
func (a kvAuth) authenticateLocalRequest(ctx context.Context, clientTenantID roachpb.TenantID) (authnResult, error) {
	if clientTenantID.IsSystem() {
		return authnSuccessPeerIsPrivileged{}, nil
	}
	
	// Sanity check: verify that we do not also have gRPC network credentials
	// in the context. This would indicate that metadata was improperly propagated.
	maybeTid, err := getTenantIDFromNetworkCredentials(ctx)
	if err != nil || maybeTid.IsSet() {
		return nil, authErrorf("programming error: network credentials in internal adapter request (%v, %v)", maybeTid, err)
	}
	return authnSuccessPeerIsTenantServer(clientTenantID), nil
}

// authenticateRemoteRequest authenticates requests made over a TLS connection.
func (a kvAuth) authenticateRemoteRequest(ctx context.Context) (authnResult, error) {
	// We will need to look at the TLS cert in any case, so
	// extract it first.
	clientCert, err := getClientCert(ctx)
	if err != nil {
		return nil, err
	}
	tenantIDFromCredentials, err := getTenantIDFromNetworkCredentials(ctx)
	if err != nil {
		return nil, authErrorf("client provided invalid tenant ID: %v", err)
	}
	
	// Did the client peer use a tenant client cert?
	if security.IsTenantCertificate(clientCert) {
		// If the peer is using a client tenant cert, in any case we
		// validate the tenant ID stored in the CN for correctness.
		tlsID, err := tenantIDFromString(clientCert.Subject.CommonName, "Common Name (CN)")
		if err != nil {
			return nil, err
		}

		// If the peer is using a TenantCertificate and also
		// provided a tenant ID via gRPC metadata, they must
		// match.
		if tenantIDFromCredentials.IsSet() && tenantIDFromCredentials != tlsID {
			return nil, authErrorf(
				"client wants to authenticate as tenant %v, but is using TLS cert for tenant %v",
				tenantIDFromCredentials, tlsID)
		}
		return authnSuccessPeerIsTenantServer(tlsID), nil
	} else {
		// We are using TLS, but the peer is not using a client tenant
		// cert. In that case, we only allow RPCs if the principal is
		// 'node' or 'root' and the tenant scope in the cert matches
		// this server (either the cert has scope "global" or its scope
		// tenant ID matches our own).
		//
		// TODO(benesch): the vast majority of RPCs should be limited to just
		// NodeUser. This is not a security concern, as RootUser has access to
		// read and write all data, merely good hygiene. For example, there is
		// no reason to permit the root user to send raw Raft RPCs.
		certUserScope, err := security.GetCertificateUserScope(clientCert)
		if err != nil {
			return nil, err
		}
		if err := checkRootOrNodeInScope(certUserScope, a.tenant.tenantID); err != nil {
			return nil, err
		}
		
		if tenantIDFromCredentials.IsSet() {
			return authnSuccessPeerIsTenantServer(tlsID), nil
		} else {
			return authnSuccessPeerIsPrivileged{}, nil
		}
	}
}
```

Reply:
Not bad. Doing this simplification in a separate PR: #96338